### PR TITLE
Adding Dockerfile for deployment option

### PR DESCRIPTION
**Pull request by prince-chrismc**
I needed to deploy an update version so I wrapped it in a docker container to deploy in my test network.

I setup automatic builds on docker hub so it always tracks master.

It would be best if there was a amwa-tv repository with this image, doing so would require the links and some commands to be updated.

### CODE REVIEW

**Comment by jasperpeeters** [comment]
Happy to see this here! I will have a closer look at this when I'm back in the office on Wednesday. I already had my own local Dockerfile but never the time to put it on here... 
In the meantime: 
@peterbrightwell as @prince-chrismc already pointed out, this should definitely be hosted by amwa. We would need an account on the official docker hub for this (hub.docker.com). Is this doable?

**Comment by peterbrightwell** [comment]
I should think this is doable (as long as it's just a public hub repo, else will cost $).  Will check Brad is happy with the terms of service and make an amwa-tv account if it's ok.

File: ReadMe.md [hunk 1]
```diff
@@ -18,15 +18,38 @@ When testing any of the above APIs it is important that they contain representat
...
+
+This image provides an quick way for deploying the test in your network.
+```shell
+docker run -d -p="5000:5000" princechrismc/nmos-testing
+```
+The web service will be available on `http://<DOCKER_HOST_IP>:5000`.
+
+If you need to change the `Config.py` settings. You can define you're own copy with the required settings and define the following volume mount.
+```shell
+docker run -d -p="5000:5000" -v="<PATH_TO_CONFIG_PY>:/config" princechrismc/nmos-testing
```
Comment (jasperpeeters) [inline ReadMe.md]: Here we could pass the path + filename in order to map the file only, not the whole folder, e.g. `-v="$(pwd)/Config.py:/config/Config.py"`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
+FROM python:3.5-alpine3.8
```
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply Dockerfile]: There was no reason in particular. It was near the top of the list when I was looking at the available tags.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+RUN apk update \
+ && apk add gcc musl-dev linux-headers git \
+ && rm -rf /var/cache/apk/*
+
+WORKDIR /home/nmos-testing
+ADD . .
+
+RUN pip3 install -r requirements.txt
+
+RUN mkdir -p /config \
```
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply Dockerfile]: That's true. I had separated them to take advantage of caching while making the Dockerfile. Good to do

File: ReadMe.md (hunk 1 above)
Comment (prince-chrismc) [reply ReadMe.md]: I did not even know single file mounts were possible! Definitely a good suggestion

**Comment by peterbrightwell** [comment]
Brad's happy for this to be hosted using and AMWA-managed docker hub, subject to checking the legal sign-off. Mary's checking now, and when done I'll help her set that up.

File: Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (prince-chrismc) [reply Dockerfile]: using python:alpine which tracks the latest of both

File: ReadMe.md [hunk 4]
```diff
@@ -20,15 +20,38 @@ When testing any of the above APIs it is important that they contain representat
...
+![](https://images.microbadger.com/badges/image/princechrismc/nmos-testing.svg)
+](https://hub.docker.com/r/princechrismc/nmos-testing)
+
+This image provides an quick way for deploying the test in your network.
+```shell
+docker run -d -p="5000:5000" princechrismc/nmos-testing
+```
+The web service will be available on `http://<DOCKER_HOST_IP>:5000`.
+
+If you need to change the `Config.py` settings. You can define you're own copy with the required settings and define the following volume mount. _Note_: use an absulute path when not working from current directory.
```
Comment (jasperpeeters) [inline ReadMe.md]: Small typo: "your" instead of "you're"

Comment (jasperpeeters) [reply ReadMe.md]: Small typo: "absolute" instead of "absulute"

**Comment by andrewbonney** [comment]
It's always a bit of a trade-off. In this case, particularly given GitHub does a decent job of alerting to any security alerts, it probably is sensible that we pin our dependencies to the highest ones that are currently available. If this isn't done as part of this PR we can always handle it separately.

**Comment by peterbrightwell** [comment]
Quick update on amwa docker hub hosting:  Mary is in discussion with Docker staff about what is required for the terms of service.  It doesn't look such a straightforward thing to tick as e.g. for Github. See https://www.docker.com/sites/default/files/d8/2019-01/docker-master-subscription-and-services-agreement-2019-01-15.pdf if you're interested.   

More updates when I get them.

**Comment by bgilmer77** [comment]
Okay – let me know when/if I need to do something.

**Comment by daniel-pebble** [comment]
Its worth noting that this dockerfile doesn't support running the BCP-003 tests as ENABLE_HTTPS isn't enabled by default.
i'm using the following setup which allows us to quickly switch between HTTPS and HTTP using docker env vars.

```
docker run --env ENABLE_HTTPS=False --name nmos-test-http -p=5080:5000 nmos-testing
docker run --env ENABLE_HTTPS=True--name nmos-test-https -p=5443:5000 nmos-testing
```


```
FROM debian:latest

RUN apt update \
 && apt install python3-pip bsdmainutils procps vim wget git dnsutils -y 

RUN git clone https://github.com/AMWA-TV/nmos-testing.git /home/nmos-testing

WORKDIR /home/nmos-testing

RUN cd testssl \
 && wget https://github.com/drwetter/testssl.sh/archive/3.0rc5.tar.gz \
 && tar -xvzf 3.0rc5.tar.gz --strip-components=1 \
 && cd .. \
 && pip3 install -r requirements.txt 

COPY entrypoint.sh /home/nmos-testing/entrypoint.sh

RUN chmod +x /home/nmos-testing/entrypoint.sh
ENTRYPOINT /home/nmos-testing/entrypoint.sh
```

```
#!/bin/bash

if [ -z "$ENABLE_HTTPS" ]
then
      echo "ENABLE_HTTPS isn't set.  Setting to False"  
	  ENABLE_HTTPS="False"
else
      echo "ENABLE_HTTPS is NOT empty"
	  if [ ${ENABLE_HTTPS,,} = "true" ]
	  then
		echo "Enable HTTP is True"
		ENABLE_HTTPS="True"
	  else
		echo "Enable HTTP is False"
		ENABLE_HTTPS="False"
	  fi
fi

sed -i "s|ENABLE_HTTPS = False|ENABLE_HTTPS = $ENABLE_HTTPS |g" Config.py

sleep 1

python3 nmos-test.py
exit $?
```

**Comment by prince-chrismc** [comment]
@daniel-pebble out curiosity did you try providing you own `Config.py` to the image of this PR? 
```
docker run -d -p="5000:5000" -v="$(pwd)/Config.py:/config/Config.py" nmos-testing
```
It's described in the ReadMe.md, and allows you to run with something other than the default settings.

Small note, I have not tested the HTTPS support from the docker image so it may be missing a dependency.

**Comment by jasperpeeters** [comment]
> Quick update on amwa docker hub hosting: Mary is in discussion with Docker staff about what is required for the terms of service. It doesn't look such a straightforward thing to tick as e.g. for Github. See https://www.docker.com/sites/default/files/d8/2019-01/docker-master-subscription-and-services-agreement-2019-01-15.pdf if you're interested.
> 
> More updates when I get them.

Not sure what the current state of the amwa docker hub hosting is, but maybe hosting it directly on github could be an option in the future: 

https://github.com/features/package-registry
https://help.github.com/en/articles/configuring-docker-for-use-with-github-package-registry

Still beta and not sure if this makes things easier (ToS wise), but sounds interesting.

**Comment by andrewbonney** [comment]
@prince-chrismc @jasperpeeters Could you confirm if this is still working when the latest changes from master are merged in? If so I think we can get this merged. We should be able to publish images to Docker Hub imminently.

**Comment by andrewbonney** [comment]
After a couple of tweaks I've got an image uploaded to https://hub.docker.com/r/amwa/nmos-testing/tags using this Dockerfile plus a few tweaks (see https://github.com/AMWA-TV/nmos-testing/tree/docker-hub-test). I'd appreciate it if anyone gets chance to try it out.

**Comment by prince-chrismc** [comment]
I verified the IS-04 in query mode and the IS-05 worked. 
Was unable to test BCP-003-1, I will try again.

**Comment by andrewbonney** [comment]
Sounds good. Is there anything outstanding from your perspective @prince-chrismc? Otherwise I'll get this merged.

**Comment by prince-chrismc** [comment]
The only extra missing is the sdpoker. Other than that everything looks
good.

**Comment by andrewbonney** [comment]
I've just added that to https://github.com/AMWA-TV/nmos-testing/tree/docker-hub-test if you'd like to merge it in here.

**Comment by prince-chrismc** [comment]
Changes looked good.

**Comment by andrewbonney** [comment]
Thanks. Sorry it took a while to get this sorted!

**Pull request merged**